mlkem/indcpa.c

```diff
@@ -145,10 +145,11 @@ void gen_matrix_entry_x4(poly *vec, uint8_t *seed[4])  // clang-format off
 // clang-format on
 {
   // Temporary buffers for XOF output before rejection sampling
-  uint8_t buf0[GEN_MATRIX_NBLOCKS * SHAKE128_RATE];
-  uint8_t buf1[GEN_MATRIX_NBLOCKS * SHAKE128_RATE];
-  uint8_t buf2[GEN_MATRIX_NBLOCKS * SHAKE128_RATE];
-  uint8_t buf3[GEN_MATRIX_NBLOCKS * SHAKE128_RATE];
+  // padding so we can load full 16-byte vectors in native implementations
+  uint8_t buf0[GEN_MATRIX_NBLOCKS * SHAKE128_RATE + 8];
```

Comment:
What is the +8 intended to achieve? Is there a particular alignment required by the AVX2 code? If so, what is it?
The required alignment should be stated as a precondition on the native API.

It might be possible to express that formally using the alignof() function. e.g.
```
REQUIRES(alignof(x) == 32)
```
in the specs of the native API.

Can you also assert here that it's correct for these declarations using a static assertion? So something like
```
STATIC_ASSERT(alignof(buf0) == 32);
```
or whatever is required.

Reply:
Actually, I think the pre-condition should be in rej_uniform.[hc] AND in the native API in arith_native.h ?

Reply:
Would it not be better to use the ALIGN macro/attribute (or similar) here?

Reply:
Sorry, had no time left to document this properly, but wanted to see if it fixes the issues asan found in the CI run that Hanno linked. It did fix it for me locally.

This is not an alignment issue.

What is required for the AVX2 code is that the buffer size is a multiple of 32 (or 16, tbd) so that full vectors can be loaded. 
What we have right now is 3*168 which needs 8 bytes of padding to be guaranteed to work.

Hanno pointed out that the number of blocks (3) may be changed at some point so 8 may be dangerous. So maybe we want to go for 24 instead - then it works for any multiple of 168.

I will fix the commit message and comment when I am back at my laptop - or if you want it merged quick, please go ahead and change it.

Alignment is a separate concern and adding a ALIGN is probably not a bad idea, although we have not seen any issues caused by this yet.

Reply:
> Alignment is a separate concern and adding a ALIGN is probably not a bad idea, although we have not seen any issues caused by this yet.

Isn't alignment a requirement, not merely a performance optimization, for AVX2?

Reply:
Yes.
```
STATIC_ASSERT(sizeof(buf0) % 32 == 0, gen_matrix_entry_x4_buf0_size_check);
```
works when -DMLKEM_DEBUG is on.

Reply:
You can also get a static assertion to check alignment:
```
  STATIC_ASSERT(alignof(buf0) == 1, gen_matrix_entry_x4_buf0_alignment_check);
```
but you have to remember to
```
#include <stdalign.h>
```
first and (strictly speaking) "alignof()" is C11 not C99...

Reply:
If it's not C99, we can't use it unfortunately. But we can just add an alignment clause just as we've done in many places already.